programmer/tinyprog/__init__.py

```diff
@@ -376,6 +379,84 @@ def program(self, addr, data):
             self.progress("Failure!")
             return False
 
+    # for each sector: erase,program,verify
+    # this programming is much more reliable than previous program()
+    # because each written sector will be verified and in case of
+    # error, retried several times
+    def program_sectors(self, addr, data, retry=10):
```

Comment:
How often does the flash write fail and need to be retried?  I haven't experienced this before.

Reply:
I have 2 USB ports and 2 types of FLASH, spansion and issi (was micron).
spansion flash works good with both ports, retry 1x now and then.
issi on one port retries very much on particular places of long uploads.
5 retries is not enough 10 is makes it work. It's not problem with the flash
but with unreliable USB implementation that looses packets.

Same binary content makes same amount of retries on exactly the same
place in the data every time, no randomness therre.

Reply:
The behavior is different between the different flash chips? If its an issue with USB I would expect the behavior to be exactly the same between the chips.  If you tried flashing the same bitstream file to both chips and observe differences in retries?

Reply:
The behaviour is differerent to different flash chips but my theory is that
this is only because different flash chips have different timings during waiting
for a sector to be erased or written and it causes USB code to be caught in
some unsupported state (bug exposed)

Same data content flashed to differrent chips behaves different in terms of retries.

But always in the same data on the same chip fail in predictable manner each time,
so my theory is that it's not e.g. worn-out flash memory or undervoltage supply which
causes data to be unpredictably written. Both chips are new, good, 100% functional and properly powered. If flashed with other tools each flash succeeds reliable and no verify
fails no retries needed.